File: ts/src/envelope.ts
export type FacileApp =
  | "sablier"
  | "opus"
  | "charles"
  | "plume"
  | "glouton"
  | "vision";

export type FacileAction = "created" | "updated" | "deleted";

export type FacileObjectType =
  | "project"
  | "task"
  | "user"
  | "time_entry"
  | "invoice"
  | "document";

export interface FacileEvent<T = unknown> {
  app: FacileApp;
  object: FacileObjectType;
  action: FacileAction;
  facile_id: string;
  payload: T;
  timestamp: string;
  idempotency_key: string;
}

File: go/envelope.go
package enveloppe

type App string

const (
	AppSablier  App = "sablier"
	AppOpus     App = "opus"
	AppCharles  App = "charles"
	AppPlume    App = "plume"
	AppGlouton  App = "glouton"
	AppVision   App = "vision"
)

type Action string

const (
	ActionCreated Action = "created"
	ActionUpdated Action = "updated"
	ActionDeleted Action = "deleted"
)

type ObjectType string

const (
	ObjectProject   ObjectType = "project"
	ObjectTask      ObjectType = "task"
	ObjectUser      ObjectType = "user"
	ObjectTimeEntry ObjectType = "time_entry"
	ObjectInvoice   ObjectType = "invoice"
	ObjectDocument  ObjectType = "document"
)

type Event[T any] struct {
	App            App        `json:"app"`
	Object         ObjectType `json:"object"`
	Action         Action     `json:"action"`
	FacileID       string     `json:"facile_id"`
	Payload        T          `json:"payload"`
	Timestamp      string     `json:"timestamp"`
	IdempotencyKey string     `json:"idempotency_key"`
}
